packages/core/src/markdown-summary.ts

```diff
@@ -0,0 +1,387 @@
+   * Cell content
+   */
+  data: string
+  /**
+   * Render cell as header
+   * (optional) default: false
+   */
+  header?: boolean
+  /**
+   * Number of columns the cell extends
+   * (optional) default: '1'
+   */
+  colspan?: string
+  /**
+   * Number of rows the cell extends
+   * (optional) default: '1'
+   */
+  rowspan?: string
+}
+
+export interface SummaryImageOptions {
+  /**
+   * The width of the image in pixels. Must be an integer without a unit.
+   * (optional)
+   */
+  width?: string
+  /**
+   * The height of the image in pixels. Must be an integer without a unit.
+   * (optional)
+   */
+  height?: string
+}
+
+class MarkdownSummary {
+  private _buffer: string
+  private _filePath?: string
+
+  constructor() {
+    this._buffer = ''
+  }
+
+  /**
+   * Finds the summary file path from the environment, rejects if env var is not found or file does not exist
+   * Also checks r/w permissions.
+   *
+   * @returns step summary file path
+   */
+  private async filePath(): Promise<string> {
+    if (this._filePath) {
+      return this._filePath
+    }
+
+    const pathFromEnv = process.env[SUMMARY_ENV_VAR]
+    if (!pathFromEnv) {
+      throw new Error(
+        `Unable to find environment variable for $${SUMMARY_ENV_VAR}. Check if your runtime environment supports markdown summaries.`
+      )
+    }
+
+    try {
+      await access(pathFromEnv, constants.R_OK | constants.W_OK)
+    } catch {
+      throw new Error(
+        `Unable to access summary file: '${pathFromEnv}'. Check if the file has correct read/write permissions.`
+      )
+    }
+
+    this._filePath = pathFromEnv
+    return this._filePath
+  }
+
+  /**
+   * Checks if the write of the current buffer will exceed the job summary upload limit
+   *
+   * @param {boolean} overwrite if the operation is overwrite (otherwise it's append)
+   *
+   * @returns {Promise<boolean>} whether or not the file will exceed the limit
+   */
+  private async willExceedLimit(overwrite: boolean): Promise<boolean> {
+    let expectedSize = 0
+    if (!overwrite) {
+      // if appending, we need to check the current size of the summary file
+      const filePath = await this.filePath()
+      const {size} = await stat(filePath)
+      expectedSize += size
+    }
+
+    const bufferLen = Buffer.byteLength(this._buffer, 'utf8')
+    expectedSize += bufferLen
+
+    return expectedSize > SUMMARY_LIMIT_BYTES
+  }
+
+  /**
+   * Wraps content in an HTML tag, adding any HTML attributes
+   *
+   * @param {string} tag HTML tag to wrap
+   * @param {string | null} content content within the tag
+   * @param {[attribute: string]: string} attrs key-value list of HTML attributes to add
+   *
+   * @returns {string} content wrapped in HTML element
+   */
+  private wrap(
+    tag: string,
+    content: string | null,
+    attrs: {[attribute: string]: string} = {}
+  ): string {
+    const htmlAttrs = Object.entries(attrs)
+      .map(([key, value]) => ` ${key}="${value}"`)
+      .join('')
```

Comment:
Do you need to `.join(' ')` with a space? Don't attributes need to separated by a space?

Reply:
On line 123 there's a space before the key, this is so that when the attributes get added there's a space between the first attribute and the tag name

Reply:
Ah I see